ARCs/arc-00xx.md

```diff
@@ -0,0 +1,484 @@
+  - Cosimo Bassi (@cusma)
+  - Adriano Di Luzio (@aldur)
+  - John Jannotti (@jannotti)
+discussions-to: <URL>
+status: Draft
+type: Standards Track
+category (*only required for Standards Track): Interface
+created: 2022-04-27
+requires: ARC-4
+---
+
+## Abstract
+
+A "Smart ASA" is an Algorand Standard Asset (ASA) controlled by a Smart Contract
+that exposes methods to create, configure, transfer, freeze, and destroy the
+asset.
+
+This ARC defines the ABI interface of such Smart Contract, the required
+metadata, and suggests a reference implementation.
+
+## Motivation
+
+The Algorand Standard Asset (ASA) is an excellent building block for on-chain
+applications. It is battle-tested and widely supported by SDKs, wallets, and
+dApps.
+
+However, the ASA lacks in flexibility and configurability. For instance, once
+issued it can't be re-configured (its unit name, decimals, maximum supply).
+Also, it is freely transferable (unless frozen). This prevents developers from
+specifying additional business logic to be checked while transferring it (think
+of royalties or [vesting](https://en.wikipedia.org/wiki/Vesting)).
+
+Enforcing transfer conditions require freezing the asset and transferring it
+through a clawback operation -- which results in a process that is opaque to
+users and wallets and a bad experience for the users.
+
+The Smart ASA defined by this ARC extends the ASA to increase its expressiveness
+and its flexibility. Developers can now introduce additional business logic
+around its operations. Wallets, dApps, and SDKs can recognize Smart ASAs and
+adjust their flows and user experiences accordingly.
+
+## Specification
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC
+2119](https://datatracker.ietf.org/doc/html/rfc2119).
+
+The following sections describe:
+
+- The ABI interface for a controlling Smart Contract (the Smart Contract that
+  controls a Smart ASA).
+- The metadata required to denote a Smart ASA and define the association between
+  an ASA and its controlling Smart Contract.
+
+### ABI Interface
+
+The ABI interface specified here draws inspiration from the transaction
+[reference](https://developer.algorand.org/docs/get-details/asa/#asset-functions)
+of an Algorand Standard Asset (ASA).
+
+To provide a unified and familiar interface between the Algorand Standard Asset
+and the Smart ASA, method names and parameters have been adapted to the ABI
+types but left otherwise unchanged.
+
+#### Asset Creation
+
+```json
+{
+  "name": "AssetCreate",
+  "args": [
+    { "type": "uint64", "name": "Total" },
+    { "type": "uint32", "name": "Decimals" },
+    { "type": "bool", "name": "DefaultFrozen" },
+    { "type": "string", "name": "UnitName" },
+    { "type": "string", "name": "AssetName" },
+    { "type": "string", "name": "URL" },
+    { "type": "[]byte", "name": "MetaDataHash" },
```

Comment:
To follow the ABI spec[^1], this needs to be changed:

```suggestion
    { "type": "byte[]", "name": "MetaDataHash" },
```

Similar for other `[]byte` usages. Otherwise PyTEAL will break while parsing the ABI JSON schema.

[^1]: https://developer.algorand.org/docs/get-details/dapps/smart-contracts/ABI/#data-types